src/mqtt_client.rs

```diff
@@ -447,6 +551,19 @@ where
             }
         }
 
+        if self.ping_response_is_overdue() {
```

Comment:
Abstract all of this logic into a `self.handle_ping()` function instead to keep this cleaner. Also, a lot of the intermediary functions that were added can likely be removed and just substituted directly into the code, since they aren't being reused.